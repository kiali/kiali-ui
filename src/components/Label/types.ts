export interface Labels {
  name: string;
  value: string;
}
