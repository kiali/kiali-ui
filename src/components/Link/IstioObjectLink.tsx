import * as React from 'react';
import { Paths } from '../../config';
import { Link } from 'react-router-dom';
import { Badge, Tooltip, TooltipPosition } from '@patternfly/react-core';
import { IstioTypes } from '../VirtualList/Config';

interface Props {
  name: string;
  namespace: string;
  type: string;
  subType?: string;
  query?: string;
}

export const GetIstioObjectUrl = (
  name: string,
  namespace: string,
  type: string,
  subType?: string,
  query?: string
): string => {
  const istioType = IstioTypes[type];
  let to = '/namespaces/' + namespace + '/' + Paths.ISTIO;

  if (!!subType) {
    to = to + '/' + istioType + '/' + subType + '/' + name;
  } else {
    to = to + '/' + istioType.url + '/' + name;
  }

  if (!!query) {
    to = to + '?' + query;
  }

  return to;
};

export class ReferenceIstioObjectLink extends React.Component<Props> {
  render() {
    const { name, namespace, type, subType } = this.props;
    const istioType = IstioTypes[type];

    return (
      <>
        <Tooltip position={TooltipPosition.top} content={<>{istioType.name}</>}>
          <Badge className={'virtualitem_badge_definition'}>{istioType.icon}</Badge>
        </Tooltip>
        <IstioObjectLink name={name} namespace={namespace} type={type} subType={subType}>
          {namespace}/{name}
        </IstioObjectLink>
      </>
    );
  }
}

class IstioObjectLink extends React.Component<Props> {
  render() {
    const { name, namespace, type, subType, query } = this.props;

    return <Link to={GetIstioObjectUrl(name, namespace, type, subType, query)}>{this.props.children}</Link>;
  }
}

export default IstioObjectLink;
