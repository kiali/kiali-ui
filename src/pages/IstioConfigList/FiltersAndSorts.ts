import { SortField } from '../../types/SortFilters';
import { IstioConfigItem } from '../../types/IstioConfigList';
import { FILTER_ACTION_APPEND, FILTER_ACTION_UPDATE, FilterType } from '../../types/Filters';

export const getType = (item: IstioConfigItem): string => {
  return item.type === 'adapter'
    ? item.type + '_' + item.adapter!.adapter
    : item.type === 'template'
    ? item.type + '_' + item.template!.template
    : item.type;
};

export const sortFields: SortField<IstioConfigItem>[] = [
  {
    id: 'namespace',
    title: 'Namespace',
    isNumeric: false,
    param: 'ns',
    compare: (a: IstioConfigItem, b: IstioConfigItem) => {
      return a.namespace.localeCompare(b.namespace) || a.name.localeCompare(b.name);
    }
  },
  {
    id: 'istiotype',
    title: 'Istio Type',
    isNumeric: false,
    param: 'it',
    compare: (a: IstioConfigItem, b: IstioConfigItem) => {
      return getType(a).localeCompare(getType(b)) || a.name.localeCompare(b.name);
    }
  },
  {
    id: 'istioname',
    title: 'Istio Name',
    isNumeric: false,
    param: 'in',
    compare: (a: IstioConfigItem, b: IstioConfigItem) => {
      // On same name order is not well defined, we need some fallback methods
      // This happens specially on adapters/templates where Istio 1.0.x calls them "handler"
      // So, we have a lot of objects with same namespace+name
      return (
        a.name.localeCompare(b.name) || a.namespace.localeCompare(b.namespace) || getType(a).localeCompare(getType(b))
      );
    }
  },
  {
    id: 'configvalidation',
    title: 'Config',
    isNumeric: false,
    param: 'cv',
    compare: (a: IstioConfigItem, b: IstioConfigItem) => {
      let sortValue = -1;
      if (a.validation && !b.validation) {
        sortValue = -1;
      } else if (!a.validation && b.validation) {
        sortValue = 1;
      } else if (!a.validation && !b.validation) {
        sortValue = 0;
      } else if (a.validation && b.validation) {
        if (a.validation.valid && !b.validation.valid) {
          sortValue = -1;
        } else if (!a.validation.valid && b.validation.valid) {
          sortValue = 1;
        } else if (a.validation.valid && b.validation.valid) {
          sortValue = a.validation.checks.length - b.validation.checks.length;
        } else if (!a.validation.valid && !b.validation.valid) {
          sortValue = b.validation.checks.length - a.validation.checks.length;
        }
      }

      return sortValue || a.name.localeCompare(b.name);
    }
  }
];

export const istioNameFilter: FilterType = {
  id: 'istioname',
  title: 'Istio Name',
  placeholder: 'Filter by Istio Name',
  filterType: 'text',
  action: FILTER_ACTION_UPDATE,
  filterValues: []
};

export const istioTypeFilter: FilterType = {
  id: 'istiotype',
  title: 'Istio Type',
  placeholder: 'Filter by Istio Type',
  filterType: 'select',
  action: FILTER_ACTION_APPEND,
  filterValues: [
    {
      id: 'Gateway',
      title: 'Gateway'
    },
    {
      id: 'VirtualService',
      title: 'VirtualService'
    },
    {
      id: 'DestinationRule',
      title: 'DestinationRule'
    },
    {
      id: 'ServiceEntry',
      title: 'ServiceEntry'
    },
    {
      id: 'Rule',
      title: 'Rule'
    },
    {
      id: 'Adapter',
      title: 'Adapter'
    },
    {
      id: 'Template',
      title: 'Template'
    },
    {
      id: 'QuotaSpec',
      title: 'QuotaSpec'
    },
    {
      id: 'QuotaSpecBinding',
      title: 'QuotaSpecBinding'
    },
    {
      id: 'Policy',
      title: 'Policy'
    },
    {
      id: 'MeshPolicy',
      title: 'MeshPolicy'
    },
    {
      id: 'ClusterRbacConfig',
      title: 'ClusterRbacConfig'
    },
    {
      id: 'RbacConfig',
      title: 'RbacConfig'
    },
    {
      id: 'Sidecar',
      title: 'Sidecar'
    },
    {
      id: 'ServiceRole',
      title: 'ServiceRole'
    },
    {
      id: 'ServiceRoleBinding',
      title: 'ServiceRoleBinding'
    }
  ]
};

export const configValidationFilter: FilterType = {
  id: 'configvalidation',
  title: 'Config',
  placeholder: 'Filter by Config Validation',
  filterType: 'select',
  action: FILTER_ACTION_APPEND,
  filterValues: [
    {
      id: 'valid',
      title: 'Valid'
    },
    {
      id: 'warning',
      title: 'Warning'
    },
    {
      id: 'notvalid',
      title: 'Not Valid'
    },
    {
      id: 'notvalidated',
      title: 'Not Validated'
    }
  ]
};

export const availableFilters: FilterType[] = [istioTypeFilter, istioNameFilter, configValidationFilter];

export const sortIstioItems = (
  unsorted: IstioConfigItem[],
  sortField: SortField<IstioConfigItem>,
  isAscending: boolean
) => {
  const sortPromise: Promise<IstioConfigItem[]> = new Promise(resolve => {
    resolve(unsorted.sort(isAscending ? sortField.compare : (a, b) => sortField.compare(b, a)));
  });

  return sortPromise;
};
