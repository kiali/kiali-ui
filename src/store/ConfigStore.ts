import { createStore, applyMiddleware, compose } from 'redux';
import { KialiAppState } from './Store';
import { persistStore, persistReducer } from 'redux-persist';
import { persistFilter } from 'redux-persist-transform-filter';
import { createTransform } from 'redux-persist';

import rootReducer from '../reducers';
import thunk from 'redux-thunk';

// defaults to localStorage for web and AsyncStorage for react-native
import storage from 'redux-persist/lib/storage';
import { INITIAL_GLOBAL_STATE } from '../reducers/GlobalState';
import { INITIAL_LOGIN_STATE } from '../reducers/LoginState';
import { INITIAL_GRAPH_STATE } from '../reducers/GraphDataState';
import { INITIAL_USER_SETTINGS_STATE } from '../reducers/UserSettingsState';
import { INITIAL_MESSAGE_CENTER_STATE } from '../reducers/MessageCenter';
import { INITIAL_STATUS_STATE } from '../reducers/HelpDropdownState';
import { INITIAL_NAMESPACE_STATE } from '../reducers/NamespaceState';
import { INITIAL_GRAFANA_STATE } from '../reducers/GrafanaState';
import { INITIAL_JAEGER_STATE } from '../reducers/JaegerState';
import { INITIAL_MESH_TLS_STATE } from '../reducers/MeshTlsState';

declare const window;

const webRoot = (window as any).WEB_ROOT ? (window as any).WEB_ROOT : undefined;
const persistKey = 'kiali-' + (webRoot && webRoot !== '/' ? webRoot.substring(1) : 'root');

// Needed to be able to whitelist fields but allowing to keep an initialState
const whitelistInputWithInitialState = (reducerName: string, inboundPaths: string[], initialState: any) =>
  createTransform(
    inboundState => persistFilter(inboundState, inboundPaths, 'whitelist'),
    outboundState => ({ ...initialState, ...outboundState }),
    { whitelist: [reducerName] }
  );

const namespacePersistFilter = whitelistInputWithInitialState(
  'namespaces',
  ['activeNamespaces'],
  INITIAL_NAMESPACE_STATE
);

const persistConfig = {
  key: persistKey,
  storage: storage,
  whitelist: ['namespaces', 'jaegerState', 'statusState'],
  transforms: [namespacePersistFilter]
};

const composeEnhancers =
  (process.env.NODE_ENV === 'development' && window && window.__REDUX_DEVTOOLS_EXTENSION_COMPOSE__) || compose;

const configureStore = (initialState: KialiAppState) => {
  // configure middlewares
  const middlewares = [thunk];
  // compose enhancers
  const enhancer = composeEnhancers(applyMiddleware(...middlewares));
  // persist reducers
  const persistentReducer = persistReducer(persistConfig, rootReducer);

  return createStore(persistentReducer, initialState, enhancer);
};

// Setup the initial state of the Redux store with defaults
// (instead of having things be undefined until they are populated by query)
// Redux 4.0 actually required this
const initialStore: KialiAppState = {
  globalState: INITIAL_GLOBAL_STATE,
  statusState: INITIAL_STATUS_STATE,
  namespaces: INITIAL_NAMESPACE_STATE,
  authentication: INITIAL_LOGIN_STATE,
  messageCenter: INITIAL_MESSAGE_CENTER_STATE,
  graph: INITIAL_GRAPH_STATE,
  userSettings: INITIAL_USER_SETTINGS_STATE,
  grafanaInfo: INITIAL_GRAFANA_STATE,
  jaegerState: INITIAL_JAEGER_STATE,
  meshTLSStatus: INITIAL_MESH_TLS_STATE
};

// pass an optional param to rehydrate state on app start
export const store = configureStore(initialStore);
export const persistor = persistStore(store);
