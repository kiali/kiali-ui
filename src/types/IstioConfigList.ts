import Namespace from './Namespace';
import {
  DestinationRule,
  DestinationRules,
  Gateway,
  IstioAdapter,
  IstioRule,
  IstioTemplate,
  ObjectValidation,
  Policy,
  QuotaSpec,
  QuotaSpecBinding,
  ServiceEntry,
  VirtualService,
  VirtualServices,
  Validations,
  RbacConfig,
  Sidecar,
  ServiceRole,
  ServiceRoleBinding,
  ClusterRbacConfig
} from './IstioObjects';
import { ResourcePermissions } from './Permissions';

export interface IstioConfigItem {
  namespace: string;
  type: string;
  name: string;
  gateway?: Gateway;
  virtualService?: VirtualService;
  destinationRule?: DestinationRule;
  serviceEntry?: ServiceEntry;
  rule?: IstioRule;
  adapter?: IstioAdapter;
  template?: IstioTemplate;
  quotaSpec?: QuotaSpec;
  quotaSpecBinding?: QuotaSpecBinding;
  policy?: Policy;
  meshPolicy?: Policy;
  clusterRbacConfig?: ClusterRbacConfig;
  rbacConfig?: RbacConfig;
  sidecar?: Sidecar;
  serviceRole?: ServiceRole;
  serviceRoleBinding?: ServiceRoleBinding;
  validation?: ObjectValidation;
}

export interface IstioConfigList {
  namespace: Namespace;
  gateways: Gateway[];
  virtualServices: VirtualServices;
  destinationRules: DestinationRules;
  serviceEntries: ServiceEntry[];
  rules: IstioRule[];
  adapters: IstioAdapter[];
  templates: IstioTemplate[];
  quotaSpecs: QuotaSpec[];
  quotaSpecBindings: QuotaSpecBinding[];
  policies: Policy[];
  meshPolicies: Policy[];
  clusterRbacConfigs: ClusterRbacConfig[];
  rbacConfigs: RbacConfig[];
  sidecars: Sidecar[];
  serviceRoles: ServiceRole[];
  serviceRoleBindings: ServiceRoleBinding[];
  permissions: { [key: string]: ResourcePermissions };
  validations: Validations;
}

export interface IstioService {
  name: string;
  namespace?: string;
  domain?: string;
  service?: string;
  labels?: { [key: string]: string };
}

export const dicIstioType = {
  Sidecar: 'sidecars',
  Gateway: 'gateways',
  VirtualService: 'virtualservices',
  DestinationRule: 'destinationrules',
  ServiceEntry: 'serviceentries',
  Rule: 'rules',
  Adapter: 'adapters',
  Template: 'templates',
  QuotaSpec: 'quotaspecs',
  QuotaSpecBinding: 'quotaspecbindings',
  Policy: 'policies',
  MeshPolicy: 'meshpolicies',
  ClusterRbacConfig: 'clusterrbacconfigs',
  RbacConfig: 'rbacconfigs',
  ServiceRole: 'serviceroles',
  ServiceRoleBinding: 'servicerolebindings',
  gateways: 'Gateway',
  virtualservices: 'VirtualService',
  destinationrules: 'DestinationRule',
  serviceentries: 'ServiceEntry',
  rules: 'Rule',
  adapters: 'Adapter',
  templates: 'Template',
  quotaspecs: 'QuotaSpec',
  quotaspecbindings: 'QuotaSpecBinding',
  instance: 'Instance',
  handler: 'Handler',
  policies: 'Policy',
  meshpolicies: 'MeshPolicy',
  clusterrbacconfigs: 'ClusterRbacConfig',
  rbacconfigs: 'RbacConfig',
  sidecars: 'Sidecar',
  serviceroles: 'ServiceRole',
  servicerolebindings: 'ServiceRoleBinding'
};

const includeName = (name: string, names: string[]) => {
  for (let i = 0; i < names.length; i++) {
    if (name.includes(names[i])) {
      return true;
    }
  }
  return false;
};

export const filterByName = (unfiltered: IstioConfigList, names: string[]): IstioConfigList => {
  if (names && names.length === 0) {
    return unfiltered;
  }
  return {
    namespace: unfiltered.namespace,
    gateways: unfiltered.gateways.filter(gw => includeName(gw.metadata.name, names)),
    virtualServices: {
      permissions: unfiltered.virtualServices.permissions,
      items: unfiltered.virtualServices.items.filter(vs => includeName(vs.metadata.name, names))
    },
    destinationRules: {
      permissions: unfiltered.destinationRules.permissions,
      items: unfiltered.destinationRules.items.filter(dr => includeName(dr.metadata.name, names))
    },
    serviceEntries: unfiltered.serviceEntries.filter(se => includeName(se.metadata.name, names)),
    rules: unfiltered.rules.filter(r => includeName(r.metadata.name, names)),
    adapters: unfiltered.adapters.filter(r => includeName(r.metadata.name, names)),
    templates: unfiltered.templates.filter(r => includeName(r.metadata.name, names)),
    quotaSpecs: unfiltered.quotaSpecs.filter(qs => includeName(qs.metadata.name, names)),
    quotaSpecBindings: unfiltered.quotaSpecBindings.filter(qsb => includeName(qsb.metadata.name, names)),
    policies: unfiltered.policies.filter(p => includeName(p.metadata.name, names)),
    meshPolicies: unfiltered.meshPolicies.filter(p => includeName(p.metadata.name, names)),
    clusterRbacConfigs: unfiltered.clusterRbacConfigs.filter(rc => includeName(rc.metadata.name, names)),
    rbacConfigs: unfiltered.rbacConfigs.filter(rc => includeName(rc.metadata.name, names)),
    sidecars: unfiltered.sidecars.filter(sc => includeName(sc.metadata.name, names)),
    serviceRoles: unfiltered.serviceRoles.filter(sr => includeName(sr.metadata.name, names)),
    serviceRoleBindings: unfiltered.serviceRoleBindings.filter(srb => includeName(srb.metadata.name, names)),
    validations: unfiltered.validations,
    permissions: unfiltered.permissions
  };
};

export const filterByConfigValidation = (unfiltered: IstioConfigItem[], configFilters: string[]): IstioConfigItem[] => {
  if (configFilters && configFilters.length === 0) {
    return unfiltered;
  }
  const filtered: IstioConfigItem[] = [];

  const filterByValid = configFilters.indexOf('Valid') > -1;
  const filterByNotValid = configFilters.indexOf('Not Valid') > -1;
  const filterByNotValidated = configFilters.indexOf('Not Validated') > -1;
  const filterByWarning = configFilters.indexOf('Warning') > -1;
  if (filterByValid && filterByNotValid && filterByNotValidated && filterByWarning) {
    return unfiltered;
  }

  unfiltered.forEach(item => {
    if (filterByValid && item.validation && item.validation.valid) {
      filtered.push(item);
    }
    if (filterByNotValid && item.validation && !item.validation.valid) {
      filtered.push(item);
    }
    if (filterByNotValidated && !item.validation) {
      filtered.push(item);
    }
    if (filterByWarning && item.validation && item.validation.checks.filter(i => i.severity === 'warning').length > 0) {
      filtered.push(item);
    }
  });
  return filtered;
};

export const toIstioItems = (istioConfigList: IstioConfigList): IstioConfigItem[] => {
  const istioItems: IstioConfigItem[] = [];

  const hasValidations = (type: string, name: string) =>
    istioConfigList.validations[type] && istioConfigList.validations[type][name];

  const nonItems = ['validations', 'permissions', 'namespace'];

  Object.keys(istioConfigList).forEach(field => {
    if (nonItems.indexOf(field) > -1) {
      // These items do not belong to the IstioConfigItem[]
      return;
    }

    const typeNameProto = dicIstioType[field.toLowerCase()]; // ex. serviceEntries -> ServiceEntry
    const typeName = typeNameProto.toLowerCase(); // ex. ServiceEntry -> serviceentry
    const entryName = typeNameProto.charAt(0).toLowerCase() + typeNameProto.slice(1);
    const itemField = istioConfigList[field];

    let entries = itemField;
    if (!(entries instanceof Array)) {
      // VirtualServices, DestinationRules
      entries = entries.items;
    }

    entries.forEach(entry => {
      const item = {
        namespace: istioConfigList.namespace.name,
        type: typeName,
        name: entry.metadata.name,
        validation: hasValidations(typeName, entry.metadata.name)
          ? istioConfigList.validations[typeName][entry.metadata.name]
          : undefined
      };

      item[entryName] = entry;
      istioItems.push(item);
    });
  });

  return istioItems;
};
