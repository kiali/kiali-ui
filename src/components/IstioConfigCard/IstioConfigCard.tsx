import * as React from 'react';
import { IstioConfigItem } from '../../types/IstioConfigList';
import { cellWidth, ICell, IRow, Table, TableBody, TableHeader, TableVariant } from '@patternfly/react-table';
import {
  Card,
  CardActions,
  CardBody,
  CardHead,
  CardHeader,
  EmptyState,
  EmptyStateBody,
  EmptyStateVariant,
  Title,
  TooltipPosition
} from '@patternfly/react-core';
import { ValidationObjectSummary } from '../Validations/ValidationObjectSummary';
import IstioObjectLink from '../Link/IstioObjectLink';
import { IstioTypes } from '../VirtualList/Config';
import { style } from 'typestyle';
import { PFBadge } from '../Pf/PfBadges';

interface Props {
  name: string;
  items: IstioConfigItem[];
}

const emtpytStyle = style({
  padding: '0 0 0 0',
  margin: '0 0 0 0'
});

class IstioConfigCard extends React.Component<Props> {
  columns(): ICell[] {
    return [{ title: 'Name' }, { title: 'Status', transforms: [cellWidth(10) as any] }];
  }

  noIstioConfig(): IRow[] {
    return [
      {
        cells: [
          {
            title: (
              <EmptyState variant={EmptyStateVariant.small} className={emtpytStyle}>
                <EmptyStateBody className={emtpytStyle}>No Istio Config found for {this.props.name}</EmptyStateBody>
              </EmptyState>
            ),
            props: { colSpan: 2 }
          }
        ]
      }
    ];
  }

  overviewLink(item: IstioConfigItem) {
    return (
      <IstioObjectLink name={item.name} namespace={item.namespace || ''} type={item.type}>
        {item.name}
      </IstioObjectLink>
    );
  }

  rows(): IRow[] {
    if (this.props.items.length === 0) {
      return this.noIstioConfig();
    }
    let rows: IRow[] = [];
    this.props.items
      .sort((a: IstioConfigItem, b: IstioConfigItem) => {
        if (a.type < b.type) {
          return -1;
        } else if (a.type > b.type) {
          return 1;
        } else {
          return a.name < b.name ? -1 : 1;
        }
      })
      .map((item, itemIdx) => {
        rows.push({
          cells: [
            {
              title: (
                <span>
                  <PFBadge badge={IstioTypes[item.type].badge} position={TooltipPosition.top} />
                  {this.overviewLink(item)}
                </span>
              )
            },
            {
              title: (
                <ValidationObjectSummary
                  id={itemIdx + '-config-validation'}
                  validations={item.validation ? [item.validation] : []}
                  style={{ verticalAlign: '-0.5em' }}
                />
              )
            }
          ]
        });
        return rows;
      });

    return rows;
  }

  render() {
    return (
      <Card isCompact={true} id={'IstioConfigCard'}>
        <CardHead>
          <CardActions />
          <CardHeader>
            <Title style={{ float: 'left' }} headingLevel="h5" size="lg">
              Istio Config
            </Title>
          </CardHeader>
        </CardHead>
        <CardBody>
          <Table
            variant={TableVariant.compact}
            aria-label={'list_istio_config'}
            cells={this.columns()}
            rows={this.rows()}
            // This style is declared on _overrides.scss
            className="table"
          >
            <TableHeader />
            <TableBody />
          </Table>
        </CardBody>
      </Card>
    );
  }
}

export default IstioConfigCard;
