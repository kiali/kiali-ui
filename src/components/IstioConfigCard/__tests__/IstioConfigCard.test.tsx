import * as React from 'react';
import { mount, shallow } from 'enzyme';
import { IstioConfigItem } from '../../../types/IstioConfigList';
import IstioConfigCard from '../IstioConfigCard';
import { shallowToJson } from 'enzyme-to-json';
import IstioObjectLink from '../../Link/IstioObjectLink';

import { store } from '../../../store/ConfigStore';
import { Provider } from 'react-redux';
import { MemoryRouter, Route } from 'react-router';

const mockComponent = (name: string, items: IstioConfigItem[]) => {
  return <IstioConfigCard name={name} items={items} />;
};

const mountComponent = (name: string, items: IstioConfigItem[]) => {
  return mount(
    <Provider store={store}>
      <MemoryRouter>
        <Route render={() => mockComponent(name, items)} />
      </MemoryRouter>
    </Provider>
  );
};

describe('IstioConfigCard', () => {
  it('shows empty state when no items', () => {
    const card = mockComponent('reviews', []);

    expect(mount(card).text()).toContain('No Istio Config found for reviews');
    expect(shallowToJson(shallow(card))).toMatchSnapshot();
  });

  it('sort by type and name', () => {
    const name = 'reviews';
    const items = [
      { type: 'virtualservice', name: 'reviews-v9', namespace: 'bookinfo' },
      { type: 'sidecar', name: 'reviews-v9', namespace: 'bookinfo' },
      { type: 'destinationrule', name: 'reviews-v2', namespace: 'bookinfo' },
      { type: 'sidecar', name: 'reviews-v1', namespace: 'bookinfo' },
      { type: 'virtualservice', name: 'reviews-v8', namespace: 'bookinfo' },
      { type: 'sidecar', name: 'reviews-v3', namespace: 'bookinfo' },
      { type: 'destinationrule', name: 'reviews-v9', namespace: 'bookinfo' },
      { type: 'destinationrule', name: 'reviews-v8', namespace: 'bookinfo' },
      { type: 'virtualservice', name: 'reviews-v5', namespace: 'bookinfo' }
    ];

    const card = mockComponent(name, items);
    const shallowed = shallow(card);
    expect(shallowToJson(shallowed)).toMatchSnapshot();

    const mounted = mountComponent(name, items);
    const links = mounted.find(IstioObjectLink);

    expect(links.at(0).props()).toEqual({
      type: 'destinationrule',
      name: 'reviews-v2',
      namespace: 'bookinfo',
      children: 'reviews-v2'
    });
    expect(links.at(1).props()).toEqual({
      type: 'destinationrule',
      name: 'reviews-v8',
      namespace: 'bookinfo',
      children: 'reviews-v8'
    });
    expect(links.at(2).props()).toEqual({
      type: 'destinationrule',
      name: 'reviews-v9',
      namespace: 'bookinfo',
      children: 'reviews-v9'
    });
    expect(links.at(3).props()).toEqual({
      type: 'sidecar',
      name: 'reviews-v1',
      namespace: 'bookinfo',
      children: 'reviews-v1'
    });
    expect(links.at(4).props()).toEqual({
      type: 'sidecar',
      name: 'reviews-v3',
      namespace: 'bookinfo',
      children: 'reviews-v3'
    });
    expect(links.at(5).props()).toEqual({
      type: 'sidecar',
      name: 'reviews-v9',
      namespace: 'bookinfo',
      children: 'reviews-v9'
    });
    expect(links.at(6).props()).toEqual({
      type: 'virtualservice',
      name: 'reviews-v5',
      namespace: 'bookinfo',
      children: 'reviews-v5'
    });
    expect(links.at(7).props()).toEqual({
      type: 'virtualservice',
      name: 'reviews-v8',
      namespace: 'bookinfo',
      children: 'reviews-v8'
    });
    expect(links.at(8).props()).toEqual({
      type: 'virtualservice',
      name: 'reviews-v9',
      namespace: 'bookinfo',
      children: 'reviews-v9'
    });

    expect(links).toHaveLength(9);
  });
});
