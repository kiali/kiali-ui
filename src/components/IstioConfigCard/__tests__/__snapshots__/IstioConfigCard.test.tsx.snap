// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`IstioConfigCard shows empty state when no items 1`] = `
<Card
  id="IstioConfigCard"
  isCompact={true}
>
  <CardHead>
    <CardActions />
    <CardHeader>
      <Title
        headingLevel="h5"
        size="lg"
        style={
          Object {
            "float": "left",
          }
        }
      >
        Istio Config
      </Title>
    </CardHeader>
  </CardHead>
  <CardBody>
    <Component
      aria-label="list_istio_config"
      cells={
        Array [
          Object {
            "title": "Name",
          },
          Object {
            "title": "Status",
            "transforms": Array [
              [Function],
            ],
          },
        ]
      }
      className="table"
      rows={
        Array [
          Object {
            "cells": Array [
              Object {
                "props": Object {
                  "colSpan": 2,
                },
                "title": <EmptyState
                  className="fvhz1n5"
                  variant="small"
                >
                  <EmptyStateBody
                    className="fvhz1n5"
                  >
                    No Istio Config found for 
                    reviews
                  </EmptyStateBody>
                </EmptyState>,
              },
            ],
          },
        ]
      }
      variant="compact"
    >
      <TableHeader />
      <TableBody />
    </Component>
  </CardBody>
</Card>
`;

exports[`IstioConfigCard sort by type and name 1`] = `
<Card
  id="IstioConfigCard"
  isCompact={true}
>
  <CardHead>
    <CardActions />
    <CardHeader>
      <Title
        headingLevel="h5"
        size="lg"
        style={
          Object {
            "float": "left",
          }
        }
      >
        Istio Config
      </Title>
    </CardHeader>
  </CardHead>
  <CardBody>
    <Component
      aria-label="list_istio_config"
      cells={
        Array [
          Object {
            "title": "Name",
          },
          Object {
            "title": "Status",
            "transforms": Array [
              [Function],
            ],
          },
        ]
      }
      className="table"
      rows={
        Array [
          Object {
            "cells": Array [
              Object {
                "title": <span>
                  <PFBadge
                    badge={
                      Object {
                        "badge": "DR",
                        "tt": "Destination Rule",
                      }
                    }
                    position="top"
                  />
                  <IstioObjectLink
                    name="reviews-v2"
                    namespace="bookinfo"
                    type="destinationrule"
                  >
                    reviews-v2
                  </IstioObjectLink>
                </span>,
              },
              Object {
                "title": <ValidationObjectSummary
                  id="0-config-validation"
                  style={
                    Object {
                      "verticalAlign": "-0.5em",
                    }
                  }
                  validations={Array []}
                />,
              },
            ],
          },
          Object {
            "cells": Array [
              Object {
                "title": <span>
                  <PFBadge
                    badge={
                      Object {
                        "badge": "DR",
                        "tt": "Destination Rule",
                      }
                    }
                    position="top"
                  />
                  <IstioObjectLink
                    name="reviews-v8"
                    namespace="bookinfo"
                    type="destinationrule"
                  >
                    reviews-v8
                  </IstioObjectLink>
                </span>,
              },
              Object {
                "title": <ValidationObjectSummary
                  id="1-config-validation"
                  style={
                    Object {
                      "verticalAlign": "-0.5em",
                    }
                  }
                  validations={Array []}
                />,
              },
            ],
          },
          Object {
            "cells": Array [
              Object {
                "title": <span>
                  <PFBadge
                    badge={
                      Object {
                        "badge": "DR",
                        "tt": "Destination Rule",
                      }
                    }
                    position="top"
                  />
                  <IstioObjectLink
                    name="reviews-v9"
                    namespace="bookinfo"
                    type="destinationrule"
                  >
                    reviews-v9
                  </IstioObjectLink>
                </span>,
              },
              Object {
                "title": <ValidationObjectSummary
                  id="2-config-validation"
                  style={
                    Object {
                      "verticalAlign": "-0.5em",
                    }
                  }
                  validations={Array []}
                />,
              },
            ],
          },
          Object {
            "cells": Array [
              Object {
                "title": <span>
                  <PFBadge
                    badge={
                      Object {
                        "badge": "SC",
                        "tt": "Istio Sidecar Proxy",
                      }
                    }
                    position="top"
                  />
                  <IstioObjectLink
                    name="reviews-v1"
                    namespace="bookinfo"
                    type="sidecar"
                  >
                    reviews-v1
                  </IstioObjectLink>
                </span>,
              },
              Object {
                "title": <ValidationObjectSummary
                  id="3-config-validation"
                  style={
                    Object {
                      "verticalAlign": "-0.5em",
                    }
                  }
                  validations={Array []}
                />,
              },
            ],
          },
          Object {
            "cells": Array [
              Object {
                "title": <span>
                  <PFBadge
                    badge={
                      Object {
                        "badge": "SC",
                        "tt": "Istio Sidecar Proxy",
                      }
                    }
                    position="top"
                  />
                  <IstioObjectLink
                    name="reviews-v3"
                    namespace="bookinfo"
                    type="sidecar"
                  >
                    reviews-v3
                  </IstioObjectLink>
                </span>,
              },
              Object {
                "title": <ValidationObjectSummary
                  id="4-config-validation"
                  style={
                    Object {
                      "verticalAlign": "-0.5em",
                    }
                  }
                  validations={Array []}
                />,
              },
            ],
          },
          Object {
            "cells": Array [
              Object {
                "title": <span>
                  <PFBadge
                    badge={
                      Object {
                        "badge": "SC",
                        "tt": "Istio Sidecar Proxy",
                      }
                    }
                    position="top"
                  />
                  <IstioObjectLink
                    name="reviews-v9"
                    namespace="bookinfo"
                    type="sidecar"
                  >
                    reviews-v9
                  </IstioObjectLink>
                </span>,
              },
              Object {
                "title": <ValidationObjectSummary
                  id="5-config-validation"
                  style={
                    Object {
                      "verticalAlign": "-0.5em",
                    }
                  }
                  validations={Array []}
                />,
              },
            ],
          },
          Object {
            "cells": Array [
              Object {
                "title": <span>
                  <PFBadge
                    badge={
                      Object {
                        "badge": "VS",
                        "tt": "Virtual Service",
                      }
                    }
                    position="top"
                  />
                  <IstioObjectLink
                    name="reviews-v5"
                    namespace="bookinfo"
                    type="virtualservice"
                  >
                    reviews-v5
                  </IstioObjectLink>
                </span>,
              },
              Object {
                "title": <ValidationObjectSummary
                  id="6-config-validation"
                  style={
                    Object {
                      "verticalAlign": "-0.5em",
                    }
                  }
                  validations={Array []}
                />,
              },
            ],
          },
          Object {
            "cells": Array [
              Object {
                "title": <span>
                  <PFBadge
                    badge={
                      Object {
                        "badge": "VS",
                        "tt": "Virtual Service",
                      }
                    }
                    position="top"
                  />
                  <IstioObjectLink
                    name="reviews-v8"
                    namespace="bookinfo"
                    type="virtualservice"
                  >
                    reviews-v8
                  </IstioObjectLink>
                </span>,
              },
              Object {
                "title": <ValidationObjectSummary
                  id="7-config-validation"
                  style={
                    Object {
                      "verticalAlign": "-0.5em",
                    }
                  }
                  validations={Array []}
                />,
              },
            ],
          },
          Object {
            "cells": Array [
              Object {
                "title": <span>
                  <PFBadge
                    badge={
                      Object {
                        "badge": "VS",
                        "tt": "Virtual Service",
                      }
                    }
                    position="top"
                  />
                  <IstioObjectLink
                    name="reviews-v9"
                    namespace="bookinfo"
                    type="virtualservice"
                  >
                    reviews-v9
                  </IstioObjectLink>
                </span>,
              },
              Object {
                "title": <ValidationObjectSummary
                  id="8-config-validation"
                  style={
                    Object {
                      "verticalAlign": "-0.5em",
                    }
                  }
                  validations={Array []}
                />,
              },
            ],
          },
        ]
      }
      variant="compact"
    >
      <TableHeader />
      <TableBody />
    </Component>
  </CardBody>
</Card>
`;
